app/renderer/components/navigation/navigator.js

```diff
@@ -0,0 +1,342 @@
+    const activeFrame = frameStateUtil.getActiveFrame(this.props.windowState)
+    const totalBlocks = activeFrame ? this.getTotalBlocks(activeFrame) : false
+    const contextMenuDetail = this.props.windowState.get('contextMenuDetail')
+    const noScriptIsVisible = this.props.windowState.getIn(['ui', 'noScriptInfo', 'isVisible'])
+    const braverySettings = siteSettings.activeSettings(this.props.activeSiteSettings, this.props.appState, appConfig)
+    const shieldEnabled = braveShieldsEnabled(activeFrame)
+
+    return <div className={cx({
+      navbarCaptionButtonContainer: true,
+      allowDragging: this.props.shouldAllowWindowDrag
+    })}>
+      <div className='navbarMenubarFlexContainer'>
+        {
+          this.props.customTitlebar.menubarVisible
+            ? <div className='menubarContainer'>
+              <Menubar
+                template={this.props.customTitlebar.menubarTemplate}
+                selectedIndex={this.props.customTitlebar.menubarSelectedIndex}
+                contextMenuSelectedIndex={this.props.customTitlebar.contextMenuSelectedIndex}
+                contextMenuDetail={contextMenuDetail}
+                autohide={getSetting(settings.AUTO_HIDE_MENU)}
+                lastFocusedSelector={this.props.customTitlebar.lastFocusedSelector} />
+              <WindowCaptionButtons windowMaximized={this.props.customTitlebar.isMaximized} />
+            </div>
+            : null
+        }
+        <div className='navigatorWrapper'
+          onDoubleClick={this.onDoubleClick}
+          onDragOver={this.onDragOver}
+          onDrop={this.onDrop}
+        >
+          <div className='backforward'>
+            <div className={cx({
+              navigationButtonContainer: true,
+              nav: true,
+              disabled: !activeTab || !activeTab.get('canGoBack') || activeTabShowingMessageBox
+            })}>
+              <LongPressButton
+                l10nId='backButton'
+                className='navigationButton backButton'
+                disabled={!activeTab || !activeTab.get('canGoBack') || activeTabShowingMessageBox}
+                onClick={this.onBack}
+                onLongPress={this.onBackLongPress}
+              />
+            </div>
+            <div className={cx({
+              navigationButtonContainer: true,
+              nav: true,
+              disabled: !activeTab || !activeTab.get('canGoForward') || activeTabShowingMessageBox
+            })}>
+              <LongPressButton
+                l10nId='forwardButton'
+                className='navigationButton forwardButton'
+                disabled={!activeTab || !activeTab.get('canGoForward') || activeTabShowingMessageBox}
+                onClick={this.onForward}
+                onLongPress={this.onForwardLongPress}
+              />
+            </div>
+          </div>
+          <NavigationBar
+            navbar={activeFrame && activeFrame.get('navbar')}
+            sites={this.props.appState.get('sites')}
```

Comment:
this is where prop lifting gets really annoying. All of these props need to be passed in to the Navigator component so they can be passed along to NavigationBar. It's very convenient to use this.props.appState instead, but unfortunately it is not good for performance